tools/spin-ec2/aws_test.go

```diff
@@ -0,0 +1,115 @@
+package main
+
+import (
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestDetectVersions_Valid(t *testing.T) {
+	instances := map[string]string{
+		"i-042ab6db828c1c318": "canary:v1.2.3:amd64:ubuntu20.04",
+		"i-042ab6db828c1c319": "canary:v1.2.3:arm64:ubuntu20.04",
+		"i-042ab6db828c1c314": "canary:v0.1.3:amd64:ubuntu20.04",
+		"i-042ab6db828c1c315": "canary:v0.1.3:arm64:ubuntu20.04",
+		"i-042ab6db828c1c316": "canary:v1.1.3:amd64:ubuntu20.04",
+		"i-042ab6db828c1c317": "canary:v1.1.3:arm64:ubuntu20.04",
+	}
+
+	expected := []string{
+		"v0.1.3",
+		"v1.1.3",
+		"v1.2.3",
+	}
+
+	actual, err := detectVersions(instances)
+	assert.NoError(t, err)
+
+	assert.Equal(t, expected, actual)
+}
+
+func TestDetectVersions_Invalid(t *testing.T) {
+	instances := map[string]string{
+		"i-042ab6db828c1c314": "canary:v0.1.a3:amd64:ubuntu20.04",
+		"i-042ab6db828c1c315": "canary:0.1.3:arm64:ubuntu20.04",
+	}
+
+	actual, err := detectVersions(instances)
+	assert.Error(t, err)
+
+	var expected []string
+	assert.Equal(t, expected, actual)
+}
+
+func TestGetVersionsToRemove(t *testing.T) {
+	actual := getVersionsToRemove([]string{
+		"v3.3.3",
+		"v2.2.2",
+		"v1.1.1",
+	})
+
+	expected := []string{
+		"v3.3.3",
```

Comment:
should not this be oldest - v1.1.1?

Reply:
the test data is misleading. In reality, the array will be sorted. I'll fix this